stan/math/rev/mat/fun/gp_exponential_cov.hpp

```diff
@@ -0,0 +1,322 @@
+ */
+template <typename T_x, typename T_s>
+class gp_exponential_cov_vari<T_x, T_s, double> : public vari {
+ public:
+  const size_t size_;
+  const size_t size_ltri_;
+  const double l_d_;
+  const double sigma_d_;
+  const double sigma_sq_d_;
+  double *dist_;
+  vari *l_vari_;
+  vari *sigma_vari_;
+  vari **cov_lower_;
+  vari **cov_diag_;
+  /**
+   * Constructor for gp_exponential_cov
+   *
+   * All memory allocated in
+   * ChainableStack's stack_alloc arena.
+   *
+   * It is critical for the efficiency of this object
+   * that the constructor create new varis that aren't
+   * popped onto the var_stack_, but rather are
+   * popped onto the var_nochain_stack_. This is
+   * controlled to the second argument to
+   * vari's constructor.
+   *
+   * @param x std::vector input that can be used in square distance
+   *    Assumes each element of x is the same size
+   * @param sigma standard deviation
+   * @param length_scale length scale
+   */
+  gp_exponential_cov_vari(const std::vector<T_x> &x, const T_s &sigma,
+                          double length_scale)
+      : vari(0.0),
+        size_(x.size()),
+        size_ltri_(size_ * (size_ - 1) / 2),
+        l_d_(value_of(length_scale)),
+        sigma_d_(value_of(sigma)),
+        sigma_sq_d_(sigma_d_ * sigma_d_),
+        dist_(ChainableStack::instance().memalloc_.alloc_array<double>(
+            size_ltri_)),
+        cov_lower_(ChainableStack::instance().memalloc_.alloc_array<vari *>(
+            size_ltri_)),
+        cov_diag_(
+            ChainableStack::instance().memalloc_.alloc_array<vari *>(size_)) {
+    double neg_inv_l = -1.0 / l_d_;
+    size_t pos = 0;
+    for (size_t j = 0; j < size_ - 1; ++j) {
+      for (size_t i = j + 1; i < size_; ++i) {
+        double dist_sq = squared_distance(x[i], x[j]);
+        dist_[pos] = dist_sq;
+        cov_lower_[pos]
+            = new vari(sigma_sq_d_ * exp(dist_sq * neg_inv_l), false);
+        ++pos;
+      }
+    }
+    for (size_t i = 0; i < size_; ++i)
+      cov_diag_[i] = new vari(sigma_sq_d_, false);
+  }
+  virtual void chain() {
+    double adjsigma = 0;
+    for (size_t i = 0; i < size_ltri_; ++i) {
+      vari *el_low = cov_lower_[i];
+      double prod_add = el_low->adj_ * el_low->val_;
+      adjsigma += prod_add;
+    }
+    for (size_t i = 0; i < size_; ++i) {
+      vari *el = cov_diag_[i];
+      adjsigma += el->adj_ * el->val_;
+    }
+    sigma_vari_->adj_ += adjsigma * 2 / sigma_d_;
+  }
+};
+/**
+ * Returns an exponential kernel.
+ *
+ * @param x std::vector input that can be used in square distance
+ *    Assumes each element of x is the same size
+ * @param sigma standard deviation
+ * @param l length scale
+ * @return squared distance
+ * @throw std::domain_error if sigma <= 0, l <= 0, or
+ *   x is nan or infinite
+ */
+template <typename T_x>
+inline typename boost::enable_if_c<
+    boost::is_same<typename scalar_type<T_x>::type, double>::value,
+    Eigen::Matrix<var, -1, -1>>::type
```

Comment:
I don't think you need the `enable_if_c` construct here. You should be able to do something like this:
```
template<>
inline var gp_exponential_cov(...)
```

See: http://en.cppreference.com/w/cpp/language/template_specialization

Reply:
whoops. The return should be `Eigen::Matrix<var, Eigen::Dynamic, Eigen::Dynamic>`